### PR TITLE
Support TypeScript 5.2 `using` / `await using` declaration

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes https://github.com/prettier/prettier/issues/15004

- Update dependencies
- Update and fix snapshots

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by sosukesuzuki** [comment]
I'll release new patch after this is merged.

File: changelog_unreleased/typescript/15321.md
Comment (fisker) [inline changelog_unreleased/typescript/15321.md:1]:
```suggestion
#### Support TypeScript 5.2 `using` / `await using` declaration (#15321 by @sosukesuzuki)

Support for the upcoming Explicit Resource Management feature in ECMAScript. [`using` / `await using` declaration](https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#using-declarations-and-explicit-resource-management)
```

File: tests/format/js/explicit-resource-management/invalid-for-using-binding-of-of.js
Comment (fisker) [inline tests/format/js/explicit-resource-management/invalid-for-using-binding-of-of.js:1]: It's valid.  

https://github.com/babel/babel/blob/main/packages/babel-parser/test/fixtures/experimental/async-explicit-resource-management/valid-for-using-binding-of-of/input.js

File: tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js
Comment (fisker) [inline tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js:3]: Why are we removing this test?

Comment (fisker) [reply tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js:3]: If it's not throwing, we can remove `typescript` parser for this test?

Comment (fisker) [reply tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js:3]: We should also ask the parser to throw.

Comment (fisker) [reply tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js:3]: https://github.com/typescript-eslint/typescript-eslint/issues/7554

File: tests/format/misc/errors/js/explicit-resource-management/invalid-for-using-binding-in.js
Comment (fisker) [inline tests/format/misc/errors/js/explicit-resource-management/invalid-for-using-binding-in.js:1]: https://github.com/typescript-eslint/typescript-eslint/issues/7555

File: tests/format/js/explicit-resource-management/invalid-for-using-binding-of-of.js
Comment (sosukesuzuki) [reply tests/format/js/explicit-resource-management/invalid-for-using-binding-of-of.js:1]: `for (await using of of of) {}` is valid and `for (using of of of) {}` is invalid?

Comment (fisker) [reply tests/format/js/explicit-resource-management/invalid-for-using-binding-of-of.js:1]: Maybe, I believe babel not making mistakes.

Comment (sosukesuzuki) [reply tests/format/js/explicit-resource-management/invalid-for-using-binding-of-of.js:1]: https://github.com/microsoft/TypeScript/issues/55555

File: tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js
Comment (sosukesuzuki) [reply tests/format/misc/errors/js/explicit-resource-management/invalid-using-binding-let.js:3]: thanks

File: tests/format/misc/errors/js/explicit-resource-management/invalid-for-using-binding-in.js.js
Comment (fisker) [inline tests/format/misc/errors/js/explicit-resource-management/invalid-for-using-binding-in.js.js:1]: File name.

File: tests/format/misc/errors/js/explicit-resource-management/invlid-using-binding-let.js
Comment (fisker) [inline tests/format/misc/errors/js/explicit-resource-management/invlid-using-binding-let.js:1]: File name. Missing  "a"

```
invalid
   ^
```

File: tests/format/js/explicit-resource-management/invalid-for-using-declaration.js
Comment (fisker) [inline tests/format/js/explicit-resource-management/invalid-for-using-declaration.js:1]: This should be valid, I think it's from babel too.

File: tests/format/js/explicit-resource-management/invalid-using-declaration-in-for.js
Comment (fisker) [inline tests/format/js/explicit-resource-management/invalid-using-declaration-in-for.js:1]: I believe these two from existing tests? They should be valid, they should be still named `”valid“`.

Better put them back, just let typescript parser to throw.

File: changelog_unreleased/typescript/15321.md
Comment (fisker) [inline changelog_unreleased/typescript/15321.md:5]:
```suggestion
```

File: tests/format/js/explicit-resource-management/jsfmt.spec.js
Comment (sosukesuzuki) [inline tests/format/js/explicit-resource-management/jsfmt.spec.js:35]: typescript-estree throws for these files (tsc doesn't throw).

**Review by fisker — APPROVED** [review]
Good job.

File: tests/format/js/explicit-resource-management/jsfmt.spec.js
Comment (fisker) [reply tests/format/js/explicit-resource-management/jsfmt.spec.js:33]: We can ignore edge cases.

**Comment by fisker** [comment]
**Size Change:** +20.6 kB (0%) 

**Total Size:** 8.5 MB

| Filename | Size | Change |
| :--- | :---: | :---: |
| `./dist/plugins/typescript.js` | 1.15 MB | +10.3 kB (+1%) |
| `./dist/plugins/typescript.mjs` | 1.15 MB | +10.3 kB (+1%) |


If it's possible, you can run

```
yarn build --file=plugins/typescript.mjs --no-minify
```

To compare the file with main branch see why the file size grows, and try to remove unnessary code.
But it's not that important.

**Comment by sosukesuzuki** [comment]
@fisker Sure! I'll merge this PR after CI is passed. And I'll try to do reduce bundle size before release.

**Comment by fisker** [comment]
If you want do it, in my experience,  I suggest you do it before merge, it's easier to switch between branches.

**Comment by sosukesuzuki** [comment]
I tried a little, but it was difficult, so I gave up once! sorry!

**Comment by fisker** [comment]
I compared the files, there is noting we can do, we are good.

**Pull request merged**
